src/ore/src/operators_async_ext.rs

```diff
@@ -166,23 +177,35 @@ impl<G: Scope> OperatorBuilderExt<G> for OperatorBuilder<G> {
                 if had_pending_notify && !scheduler.is_notified() {
                     waker.wake_by_ref();
                 }
+
+                *reschedule_flag.borrow()
             }
         });
     }
 }
 
-#[allow(unused_macros)]
+/// Convenience macro used to wrap what might otherwise be the argument to `build_reschedule`.
+#[macro_export]
 macro_rules! async_op {
     (|$capabilities:ident, $frontiers:ident| $body:block) => {
-        move |mut capabilities, mut frontiers, scheduler| async move {
+        move |mut capabilities, mut frontiers, scheduler, reschedule_flag| async move {
             loop {
                 scheduler.notified().await;
                 // rebind to mutable references to make sure they can't be accidentally dropped
                 #[allow(unused_mut)]
                 let mut $capabilities = &mut capabilities;
                 #[allow(unused_mut)]
                 let mut $frontiers = &mut frontiers;
-                let _: () = async { $body }.await;
+
+                {
+                    // If this async block is `poll`ed and returns while `$body` is executing, it
+                    // should always be rescheduled.
+                    *reschedule_flag.borrow_mut() = true
+                }
+                let result: bool = async { $body }.await;
```

Comment:
Given the comment above here we can still require the inner loop future passed to the `async_op!` macro to return a bool indicating if it's done and use that to break out of the `loop`, which will in turn resolve the outer future and cause `build_async` to return false to `build_reschedule`